config/src/main/java/com/typesafe/config/impl/ConfigParser.java

```diff
@@ -0,0 +1,435 @@
+/**
+ *   Copyright (C) 2011-2012 Typesafe Inc. <http://typesafe.com>
+ */
+package com.typesafe.config.impl;
+
+import java.io.File;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.ListIterator;
+import java.util.Map;
+import java.util.Stack;
+
+import com.typesafe.config.*;
+
+final class ConfigParser {
+    static AbstractConfigValue parse(ConfigNodeRoot document,
+                                     ConfigOrigin origin, ConfigParseOptions options,
+                                     ConfigIncludeContext includeContext) {
+        ParseContext context = new ParseContext(options.getSyntax(), origin, document,
+                SimpleIncluder.makeFull(options.getIncluder()), includeContext);
+        return context.parse();
+    }
+
+    static private final class ParseContext {
+        private int lineNumber;
+        final private ConfigNodeRoot document;
+        final private FullIncluder includer;
+        final private ConfigIncludeContext includeContext;
+        final private ConfigSyntax flavor;
+        final private ConfigOrigin baseOrigin;
+        final private LinkedList<Path> pathStack;
+        // this is the number of "equals" we are inside,
+        // used to modify the error message to reflect that
+        // someone may think this is .properties format.
+        int equalsCount;
+        // the number of lists we are inside; this is used to detect the "cannot
+        // generate a reference to a list element" problem, and once we fix that
+        // problem we should be able to get rid of this variable.
+        int arrayCount;
```

Comment:
this feels like we have several things replicated between the two parsers; that feels like some kind of separation-of-concerns problem, but I have to read on before I have a constructive suggestion :-)